### ISSUE TITLE
RethrowCaughtException should not be triggered when there is more general exception below

### ISSUE THREAD

**Issue by matejdro** [detekt/detekt#4367]
## Expected Behavior of the rule

Following code should be compliant with RethrowCaughtException rule:

```kotlin
try {
    // Do something
} catch (e: CancellationException) {
    // Just throw cancellation
    throw e
} catch (e: Exception) {
    // Log everything except cancellation
    logException(e)
}
```

## Context

In above case I want to log all exceptions except for the Cancellation exception (which I just want to propagate). That's why CancellationException has its own catch that gets activated before the general one, which is then never reached. This cause clause is not useless because it is used to filter more general cause clauses that come after it.

Above snippet is just a more cleaner way to write following code, which is compliant:

```kotlin
try {
    // Do something
} catch (e: Exception) {
    if (e is CancellationException) {
        throw e
    } else {
        logException(e)
    }
}
```

**Comment by BraisGabin** [comment]
You are right. Do you want to give it a try and send a PR solving it?

**Comment by strkkk** [comment]
@BraisGabin I would like to give it a try

**Comment by BraisGabin** [comment]
Go for it! Let me know if you need any help :)
